### PR TITLE
plat/kvm_x86 : adding ASLR to Unikraft - updated

### PR DESCRIPTION
**Pull request by SerbanSo**
### Prerequisite checklist

 - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): x86_64
 - Platform(s): kvm
 - Application: code base and python3/3.10

### Additional configuration

 - CONFIG_LINK_ASLR=y
 - CONFIG_RUNTIME_ASLR=y
(the last requires CONFIG_LIBUKSWRAND)

### Description of changes

Fixed ASLR library and stack/heap randomization to be compatible with the latest version on the staging branch.

### CODE REVIEW

File: lib/ukboot/boot.c [hunk 1]
```diff
@@ -87,6 +87,9 @@
 #include <uk/sp.h>
 #endif
 #include <uk/arch/paging.h>
+#ifdef CONFIG_RUNTIME_ASLR
+#include <uk/swrand.h>
+#endif
```
Comment (Alex-deVis) [inline lib/ukboot/boot.c:92]:
```suggestion
#endif /* CONFIG_RUNTIME_ASLR */
```

Comment (Alex-deVis) [inline lib/ukboot/boot.c:92]:
```suggestion
#endif /* CONFIG_RUNTIME_ASLR */
```

File: plat/kvm/x86/setup.c [hunk 2]
```diff
@@ -18,12 +18,199 @@
 #include <uk/intctlr.h>
 
 #include <kvm/console.h>
+#ifdef CONFIG_RUNTIME_ASLR
+#include <uk/swrand.h>
+#endif
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:23]:
```suggestion
#endif /* CONFIG_RUNTIME_ASLR */
```

File: plat/kvm/x86/setup.c [hunk 3]
```diff
@@ -18,12 +18,199 @@
...
+
+			return NULL;
+		}
+
+		return (void *)pstart;
+	}
+
+	return NULL;
+}
+#endif
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:212]:
```suggestion
#endif /* CONFIG_RUNTIME_ASLR */
```

File: plat/kvm/x86/setup.c [hunk 4]
```diff
@@ -69,7 +256,32 @@ static void __noreturn _ukplat_entry2(void)
...
+	uk_pr_info("Initialize random number generator...\n");
 
+	for (i = 0; i < seedc; i++)
+		seedv[i] = _gen_seed32();
+
+	uk_swrand_init_r(&uk_swrand_def, seedc, seedv);
+
+	return seedc;
+}
+#endif
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:284]:
```suggestion
#endif /* CONFIG_RUNTIME_ASLR */
```

File: plat/kvm/x86/setup.c [hunk 5]
```diff
@@ -69,7 +256,32 @@ static void __noreturn _ukplat_entry2(void)
...
+static int uk_swrand_init(void)
+{
+	unsigned int i;
+#ifdef CONFIG_LIBUKSWRAND_CHACHA
+	unsigned int seedc = 10;
+	__u32 seedv[10];
+#else
+	unsigned int seedc = 2;
+	__u32 seedv[2];
+#endif
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:274]:
```suggestion
#endif /* CONFIG_LIBUKSWRAND_CHACHA */
```

File: plat/kvm/x86/setup.c [hunk 6]
```diff
@@ -69,7 +256,32 @@ static void __noreturn _ukplat_entry2(void)
 
 	ukplat_lcpu_halt();
 }
+#ifdef CONFIG_RUNTIME_ASLR
+__u32 _gen_seed32(){
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:260]: What is the point of redefining `uk_swrandr_gen_seed32()`? Also, on a closer look, it looks identical to `rdtsc()` from `plat/common/include/x86/cpu.h`.

File: plat/kvm/x86/setup.c [hunk 7]
```diff
@@ -69,7 +256,32 @@ static void __noreturn _ukplat_entry2(void)
...
+	__u32 low,high;
+	__asm__ __volatile__ ("rdtsc" : "=a" (low), "=d" (high));
+	return ((unsigned long long)high << 32) | low;
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:263]:
```suggestion
	return ((__u64)high << 32) | low;
```
Change the function signature to return a `__u64` too.

File: plat/kvm/x86/setup.c [hunk 8]
```diff
@@ -95,11 +307,17 @@ void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi)
...
-	bstack = ukplat_memregion_alloc(__STACK_SIZE, UKPLAT_MEMRT_STACK,
-					UKPLAT_MEMRF_READ |
-					UKPLAT_MEMRF_WRITE |
-					UKPLAT_MEMRF_MAP);
+	ukplat_bootinfo_print();
+	#ifdef CONFIG_RUNTIME_ASLR
+	bstack = stackmemory_aslr_palloc(__STACK_SIZE);
+	#else
+	bstack = bootmemory_palloc(__STACK_SIZE, UKPLAT_MEMRT_STACK);
+	#endif
```
Comment (Alex-deVis) [inline plat/kvm/x86/setup.c:320]:
```suggestion
	#endif /* CONFIG_RUNTIME_ASLR */
```

**Review by Alex-deVis — COMMENTED** [review]
I left a few inline comments. I did not understand why you redefine functions from `ukswrand` and call its init yourself. Assuming the reason is the weak seed used by `ukswrand` by default (i.e. timestamp), this should be solved by the [ongoing PR](https://github.com/unikraft/unikraft/pull/1008/files) to refactor the library.

**Comment by SerbanSo** [comment]
Please see: https://github.com/unikraft/unikraft/pull/1317

**Pull request closed without merging**